bittensor/utils/async_substrate.py

```diff
@@ -138,142 +649,179 @@ async def _process_response(
                             decoded_response, complete = await self._process_response(
-                                response, value_scale_type, storage_item, metadata
+                                response,
+                                value_scale_type,
+                                storage_item,
+                                runtime,
+                                result_handler,
                             )
-                            await request_manager.add_response(
+                            request_manager.add_response(
                                 item_id, decoded_response, complete
                             )
 
-                if await request_manager.is_complete():
+                if request_manager.is_complete:
                     break
 
-        return await request_manager.get_results()
+        return request_manager.get_results()
+
+    @staticmethod
+    def make_payload(id_: str, method: str, params: list) -> dict:
+        return {
+            "id": id_,
+            "payload": {"jsonrpc": "2.0", "method": method, "params": params},
+        }
 
-    @ensure_initialized
     async def rpc_request(
-        self, method: str, params: list, block_hash: Optional[str] = None
+        self,
+        method: str,
+        params: list,
+        block_hash: Optional[str] = None,
+        reuse_block_hash: bool = False,
     ) -> Any:
+        """
+        Makes an RPC request to the subtensor. Use this only if ``self.query`` and ``self.query_multiple`` and
+        ``self.query_map`` do not meet your needs.
+        """
+        block_hash = self._get_current_block_hash(block_hash, reuse_block_hash)
         payloads = [
-            {
-                "id": "rpc_request",
-                "payload": {
-                    "jsonrpc": "2.0",
-                    "method": method,
-                    "params": params + [block_hash] if block_hash else params,
-                },
-            }
+            self.make_payload(
+                "rpc_request",
+                method,
+                params + [block_hash] if block_hash else params,
+            )
         ]
-        result = await self._make_rpc_request(payloads)
+        runtime = Runtime(
+            self.chain, self.substrate.runtime_config, self.substrate.metadata
+        )
+        result = await self._make_rpc_request(payloads, runtime=runtime)
         if "error" in result["rpc_request"][0]:
             raise SubstrateRequestException(
                 result["rpc_request"][0]["error"]["message"]
             )
         return result["rpc_request"][0]
 
-    async def get_block_hash(self, block_id: int) -> str:
-        response = await self.rpc_request("chain_getBlockHash", [block_id])
-
-        if "error" in response:
-            raise SubstrateRequestException(response["error"]["message"])
-        else:
-            return response.get("result")
+    async def get_block_hash(self, block: int) -> str:
```

Comment:
Could you pls change `block` to the `block_id`? this is gonna be more obvious. The same as in original method.
Pls add docstring